source/common/grpc/buffered_async_client.h

```diff
@@ -0,0 +1,114 @@
+#pragma once
+
+#include "source/common/grpc/typed_async_client.h"
+#include "source/common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Grpc {
+
+enum class BufferState { Buffered, PendingFlush };
+
+template <class RequestType, class ResponseType> class BufferedAsyncClient {
+public:
+  BufferedAsyncClient(uint32_t max_buffer_bytes, const Protobuf::MethodDescriptor& service_method,
+                      Grpc::AsyncStreamCallbacks<ResponseType>& callbacks,
+                      const Grpc::AsyncClient<RequestType, ResponseType>& client)
+      : max_buffer_bytes_(max_buffer_bytes), service_method_(service_method), callbacks_(callbacks),
+        client_(client) {}
+
+  virtual ~BufferedAsyncClient() { cleanup(); }
+
+  uint32_t publishId(RequestType& message) { return MessageUtil::hash(message); }
+
+  void bufferMessage(uint32_t id, RequestType& message) {
+    const auto buffer_size = message.ByteSizeLong();
+    if (current_buffer_bytes_ + buffer_size > max_buffer_bytes_) {
+      return;
+    }
+
+    message_buffer_[id] = std::make_pair(BufferState::Buffered, message);
+    current_buffer_bytes_ += buffer_size;
+  }
+
+  absl::flat_hash_set<uint32_t> sendBufferedMessages() {
+    if (active_stream_ == nullptr) {
+      active_stream_ =
+          client_.start(service_method_, callbacks_, Http::AsyncClient::StreamOptions());
+    }
+
+    if (active_stream_->isAboveWriteBufferHighWatermark()) {
+      return {};
+    }
+
+    absl::flat_hash_set<uint32_t> inflight_message_ids;
+
+    for (auto&& it : message_buffer_) {
+      const auto id = it.first;
+      auto& state = it.second.first;
+      auto& message = it.second.second;
+
+      if (state == BufferState::PendingFlush) {
+        continue;
+      }
+
+      state = BufferState::PendingFlush;
+      inflight_message_ids.emplace(id);
+      active_stream_->sendMessage(message, false);
+    }
```

Comment:
This shuffles the order of the messages that the client gets, and I don't think it's a desired behavior, thoughts?

Reply:
I think that we don't have to take care about the order of received messages. It is because we can track unordered messages with a specific ID.

Reply:
No we should, otherwise there could be some case that it would leave a message in buffer infinitely.

Reply:
does it mean we should ordered tree map like absl::btree_map?

Reply:
you might need a linked hash map like structure here (using std::list and map<int, list::iterator>) to achieve that. when we have order the timeout shouldn't be an issue.

Reply:
Changed to an ordered data structure. Also, by issuing thread local IDs, we can send messages that the old stream could not send even after the active stream is closed and a new stream is issued.